validation/elliptic_solvers/doubly_bounded_poisson.jl

```diff
@@ -0,0 +1,194 @@
+using Oceananigans
+using Oceananigans.Architectures: child_architecture, device
+using Oceananigans.Operators: volume, Δyᶠᶜᵃ, Δyᶜᶠᵃ, Δyᶜᶜᵃ, Δxᶠᶜᵃ, Δxᶜᶠᵃ, Δxᶜᶜᵃ, Δyᵃᶜᵃ, Δxᶜᵃᵃ, Δzᵃᵃᶠ, Δzᵃᵃᶜ, ∇²ᶜᶜᶜ
+using KernelAbstractions: @kernel, @index, Event
+using Oceananigans.Utils: launch!
+using Oceananigans.BoundaryConditions: fill_halo_regions!
+using Oceananigans.Solvers: FFTBasedPoissonSolver, solve!, HeptadiagonalIterativeSolver, constructors, arch_sparse_matrix, matrix_from_coefficients, PreconditionedConjugateGradientSolver, MultigridSolver
+using Oceananigans.Architectures: architecture, arch_array
+using Statistics: mean
+using IterativeSolvers
+using Statistics: mean
+using AlgebraicMultigrid
+using GLMakie
+using AlgebraicMultigrid: _solve!
+using SparseArrays
+using OffsetArrays
+
+import Oceananigans.Solvers: precondition!
+
+N = 64
+
+grid = RectilinearGrid(size=(N, N), x=(-4, 4), y=(-4, 4), topology=(Bounded, Bounded, Flat))
+
+arch = architecture(grid)
+Nx, Ny, Nz = grid.Nx, grid.Ny, grid.Nz
+
+# Select RHS
+r = CenterField(grid)
+r₀(x, y, z) = (x^2 + y^2) < 1 ? 1 : 0 #exp(-x^2 - y^2)
+set!(r, r₀)
+r .-= mean(r)
+fill_halo_regions!(r, grid.architecture)
+
+
+@kernel function ∇²!(∇²f, grid, f)
+    i, j, k = @index(Global, NTuple)
+    @inbounds ∇²f[i, j, k] = ∇²ᶜᶜᶜ(i, j, k, grid, f)
+end
+
+function compute_∇²!(∇²φ, φ, arch, grid)
+    fill_halo_regions!(φ)
+    child_arch = child_architecture(arch)
+    event = launch!(child_arch, grid, :xyz, ∇²!, ∇²φ, grid, φ, dependencies=Event(device(child_arch)))
+    wait(device(child_arch), event)
+    fill_halo_regions!(∇²φ)
+
+    return nothing
+end
+
+# Solve ∇²φ = r with `FFTBasedPoissonSolver`
+φ_fft = CenterField(grid)
+fft_solver = FFTBasedPoissonSolver(grid)
+fft_solver.storage .= interior(r)
+
+@info "Solving the Poisson equation with an FFT-based solver..."
+@time solve!(φ_fft, fft_solver, fft_solver.storage)
+
+fill_halo_regions!(φ_fft)
+
+
+# Solve ∇²φ = r with `PreconditionedConjugateGradientSolver`
+φ_cg = CenterField(grid)
+cg_solver = PreconditionedConjugateGradientSolver(compute_∇²!, template_field=r, reltol=eps(eltype(grid)))
+
+@info "Solving the Poisson equation with a conjugate gradient iterative solver..."
+@time solve!(φ_cg, cg_solver, r, arch, grid)
+
+fill_halo_regions!(φ_cg)
+
+
+# Solve ∇²φ = r with `AlgebraicMultigrid` solver
+function create_matrix(grid, linear_operator!, args...)
+    Nx, Ny, Nz = grid.Nx, grid.Ny, grid.Nz
+    # Currently assuming Nz = 1
+    A = spzeros(Float64, Nx*Ny, Nx*Ny)
```

Comment:
```suggestion
    A = spzeros(eltype(grid), Nx*Ny, Nx*Ny)
```

Reply:
you can simply click "Commit suggestion" and it'll go though! Then you have to do `git pull` to update your local version...